### PR TITLE
Sdcicd 1052 added osde2e test harness

### PR DESCRIPTION
**Pull request by ritmun**
### What type of PR is this?
e2e testing

### What this PR does / why we need it?
1. Adds osde2e bp convention
2. creates a e2e test harness within MUO repo. 

The test harness contains e2e tests currently running within osde2e framework. Please review the PR to ensure it looks good and no harm is done to the operator.
Please note that this is just the first phase of decentralizing your e2e testing. 

### Which Jira/Github issue(s) this PR fixes?
[SDCICD-1052](https://issues.redhat.com//browse/SDCICD-1052)
 

### Special notes for your reviewer:
We’ll continue to run the existing tests within osde2e, and are NOT changing the osde2e pipeline to run this yet. That will be planned in Q3. With that said, any new e2e tests for MUO should **not** be added to the osde2e repo after this merge, since this harness will be the new location for the e2e tests.

Based on my thread here, I've removed the spec for UC with past start date https://redhat-internal.slack.com/archives/CFJD1NZFT/p1688076884075809 

 
To test this harness locally, comment suiteConfig, reporterConfig from osde2e/managed_upgrade_operator_runner_test.go  and run `KUBECONFIG=(your kubeadm config) OCM_ENVIRONMENT=stage ginkgo run --tags osde2e ./osde2e/ -v`

### Pre-checks (if applicable):
- [x] Tested latest changes against a cluster
- [ ] Included documentation changes with PR

### Post checks
app-interface PR to add harness image publishing to operator build
followup task: MErge this PR  https://gitlab.cee.redhat.com/service/app-interface/-/merge_requests/72934

### CODE REVIEW

File: osde2e/managed_upgrade_operator_tests.go [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+
+var _ = ginkgo.Describe("managed-upgrade-operator", ginkgo.Ordered, func() {
+	ginkgo.BeforeAll(func(ctx context.Context) {
+		log.SetLogger(ginkgo.GinkgoLogr)
+		var err error
+		k8s, err = openshift.New(ginkgo.GinkgoLogr)
+		Expect(err).ShouldNot(HaveOccurred(), "Unable to setup k8s client")
+		Expect(upgradev1alpha1.AddToScheme(k8s.GetScheme())).Should(BeNil(), "Unable to register upgradev1alpha1 api scheme")
+		impersonatedResourceClient, err = k8s.Impersonate("test-user@redhat.com", "dedicated-admins")
+		Expect(err).ShouldNot(HaveOccurred(), "Unable to setup k8s client")
```
Comment (ravitri) [inline osde2e/managed_upgrade_operator_tests.go:55]:
```suggestion
		Expect(err).ShouldNot(HaveOccurred(), "Unable to setup impersonated k8s client")
```

Comment (ravitri) [reply osde2e/managed_upgrade_operator_tests.go:55]: Line 52 and 55 report same error message but have different conditions so thought of rewording it to distinguish.

File: osde2e/managed_upgrade_operator_tests.go [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+	upgradeConfig              upgradev1alpha1.UpgradeConfig
+)
+
+var _ = ginkgo.Describe("managed-upgrade-operator", ginkgo.Ordered, func() {
+	ginkgo.BeforeAll(func(ctx context.Context) {
+		log.SetLogger(ginkgo.GinkgoLogr)
+		var err error
+		k8s, err = openshift.New(ginkgo.GinkgoLogr)
+		Expect(err).ShouldNot(HaveOccurred(), "Unable to setup k8s client")
+		Expect(upgradev1alpha1.AddToScheme(k8s.GetScheme())).Should(BeNil(), "Unable to register upgradev1alpha1 api scheme")
```
Comment (ravitri) [inline osde2e/managed_upgrade_operator_tests.go:53]:
```suggestion
		Expect(upgradev1alpha1.AddToScheme(k8s.GetScheme())).Should(BeNil(), "Unable to register upgradev1alpha1 api to openshift client scheme")
```

File: osde2e/managed_upgrade_operator_tests.go [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+		impersonatedResourceClient, err = k8s.Impersonate("test-user@redhat.com", "dedicated-admins")
+		Expect(err).ShouldNot(HaveOccurred(), "Unable to setup k8s client")
+		Expect(upgradev1alpha1.AddToScheme(impersonatedResourceClient.GetScheme())).Should(BeNil(), "Unable to register upgradev1alpha1 api scheme")
```
Comment (ravitri) [inline osde2e/managed_upgrade_operator_tests.go:56]:
```suggestion
		Expect(upgradev1alpha1.AddToScheme(impersonatedResourceClient.GetScheme())).Should(BeNil(), "Unable to register upgradev1alpha1 api to impersonated client scheme")
```

File: osde2e/managed_upgrade_operator_tests.go (hunk 2 above)
Comment (ravitri) [reply osde2e/managed_upgrade_operator_tests.go:53]: Differentiating error message in line 53 and 55 for failure to register api scheme

File: osde2e/managed_upgrade_operator_tests.go (hunk 3 above)
Comment (ravitri) [reply osde2e/managed_upgrade_operator_tests.go:56]: Differentiating error message in line 53 and 55 for failure to register api scheme

File: osde2e/managed_upgrade_operator_tests.go [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+		})
+
+		ginkgo.It("should raise prometheus metric if start time is invalid", func(ctx context.Context) {
+			skipIfNoUpdatesOrCurrentlyUpgrading(ctx, clusterVersion)
+			targetVersion := clusterVersion.Status.AvailableUpdates[0].Version
+			targetChannel := clusterVersion.Spec.Channel
+			upgradeConfig = makeUpgradeConfig(upgradeConfigResourceName, operatorNamespace, "this is not a start time", targetVersion, targetChannel)
+
+			ginkgo.By("Creating test upgrade config")
+			err = k8s.WithNamespace(operatorNamespace).Create(ctx, &upgradeConfig)
```
Comment (jbpratt) [inline osde2e/managed_upgrade_operator_tests.go:150]:
```suggestion
			err = k8s.Create(ctx, &upgradeConfig)
```

File: osde2e/managed_upgrade_operator_tests.go (hunk 2 above)
Comment (jbpratt) [inline osde2e/managed_upgrade_operator_tests.go:53]:
```suggestion
		Expect(upgradev1alpha1.AddToScheme(k8s.GetScheme())).Should(Succeed(), "Unable to register upgradev1alpha1 api scheme")
```

to make the assertion a bit more clear, we can be specific when we use `Expect(thing()).Should(BeNil())` vs something like `Expect(thing()).Should(Succeed())`

this goes for a lot of the assertions below as well

File: osde2e/managed_upgrade_operator_tests.go [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const (
+	operatorName                           = "managed-upgrade-operator"
+	operatorNamespace                      = "openshift-managed-upgrade-operator"
+	operatorLockFile                       = "managed-upgrade-operator-lock"
+	upgradeConfigResourceName              = "managed-upgrade-config"
+	upgradeConfigForDedicatedAdminTestName = "osde2e-da-upgrade-config"
+	defaultDesiredReplicas                 = 1
```
Comment (jbpratt) [inline osde2e/managed_upgrade_operator_tests.go:34]: unused var (if we scope these lower into the `Describe` then we will start seeing the compiler complain about this)

**Pull request merged**